src/lib/cfa/sudoers.rb

```diff
@@ -0,0 +1,237 @@
+        end
+        spec["run_as"] = values["host_group"]["command[]"]["runas_user"]
+        spec["tag"] = values["host_group"]["command[]"]["tag"]
+
+        specs << spec.reject { |_, v| v.nil? }
+      end
+    end
+
+    # Remove all users specifications
+    def clean_users
+      data.delete(USERS_MATCHER)
+    end
+
+    # Add a new basic user specification
+    #
+    #
+    # A user specification determines which commands a user may run (and as what user) on specified
+    # hosts (who may run what, for short).
+    #
+    # Its basic structure is “who where = (as_whom) what”.
+    #
+    # E.g., with
+    #
+    #   root            ALL = (ALL) ALL
+    #   %wheel          ALL = (ALL) ALL
+    #
+    #   We let root and any user in group wheel run any command on any host as any user.
+    #
+    # See the sudoers manpage for more details.
+    #
+    # @param user_spec [Hash] a user specification
+    # @option user_spec [String] "user" who can run the command
+    # @option user_spec [String] "host" where the user can run the command
+    # @option user_spec [Array<String>] "commands" commands that the user can run
+    # @option user_spec [String] "run_as" as whom the commands will be run
+    # @option user_spec [String] "tag" the tag associated to the command (e.g. NOPASSWD)
+    def add_user(user_spec)
+      user = AugeasTree.new
+      host_group = AugeasTree.new
+
+      host_group.add("host", user_spec["host"])
+
+      commands = user_spec["commands"].reduce([]) do |commands, command|
+        augeas_command = AugeasTree.new
+        host_group.add("command[]", AugeasTreeValue.new(augeas_command, command))
+        commands << augeas_command
+      end
+
+      # For the time being, "runas_user" and "tag" options are added only to the first command,
+      # which means that they apply to all commands.
+      run_as = user_spec["run_as"].to_s.strip
+      tag = user_spec["tag"].to_s.strip
+      commands[0].add("runas_user", run_as) unless run_as.empty?
+      commands[0].add("tag", tag) unless tag.empty?
+
+      user.add("user", user_spec["user"])
+      user.add("host_group", host_group)
+
+      data.add("spec[]", user, placer)
+    end
+
+    private
+
+    attr_accessor :start_placer
+
+    # Return the current users collection
+    #
+    # @return [Array<Hash>]
+    def current_users
+      data.select(USERS_MATCHER)
+    end
+
+    COLLECTION_KEYS = ["spec", "command"].freeze
+    private_constant :COLLECTION_KEYS
+
+    # Fix collection names adding the missing "[]"
+    #
+    # When a collection has only one element, Augeas does not add "[]" suffix, reason why is
+    # necessary to fix at least those collextions that are going to be hitted for either, read or
+    # write.
+    #
+    # @param tree [AugeasTree,AugeasTreeValue]
+    def fix_collection_names(tree)
+      tree.data.each do |entry|
+        key, value = entry.values_at(:key, :value)
+
+        entry[:key] << "[]" if COLLECTION_KEYS.include?(key)
```

Comment:
this does not work as it uses data that is just copy and also frozen copy `https://github.com/config-files-api/config_files_api/blob/master/lib/cfa/augeas_parser.rb#L212` What you need to do is "all_data", but you have to be careful to not modify any data that is marked for removal. But ideally you should use just simple `select` that select given data that needs to be converted. This needs to be for sure tested properly :)

Reply:
> this does not work

:thinking: :thinking: :thinking: from my tests, I think it is working. Even more, I took this borrowed from [CFA::ChronyConf](https://github.com/yast/yast-ntp-client/blob/3b6290a437566f1c6e57ede0fd4d0deaddccac75/src/lib/cfa/chrony_conf.rb#L125-L137) model.

Shall we discard this approach?

Reply:
well, data should be frozen. So there are two options:

1. freeze is not recursive or properly implemented for augeas
2. you have older version of cfa where it is not yet frozen

both does not look for me reliable for future.

Reply:
It seems that my CFA version is up to date (1.0.1).

So, should we look for a different approach to solve this issue with the collection names?